### ISSUE TITLE
includeAll is slow

### ISSUE THREAD

**Issue by kevinburkeshyp** [balderdashy/include-all#10]
loading 40 models currently takes ~2.5 seconds on my MBP. a quick read of the source code here reveals that after files are statted, there's a call to `files.forEach`, which will require each source module in turn.

switching forEach to a non-blocking function call would be much faster. i'll probably see if i can work on a PR for this.

**Comment by kevinburkeshyp** [comment]
Might actually not understand very well what about `require` is being slow/ would benefit from an async speedup. Is there any way for one process to simultaneously run multiple `require` statements?

One `require` on my system is taking 700ms.

**Comment by kevinburkeshyp** [comment]
Wrote up more info here. https://kev.inburke.com/kevin/node-require-is-dog-slow/

**Comment by tjwebb** [comment]
@kevinburkeshyp still interested in working on this at all?

**Comment by kevinburkeshyp** [comment]
yeah, not sure there's much room for improvement... didn't realize `require` is a sync call/ there's no way to do multiple of them at one time. this sped up require times on sails by ~15-20%: https://github.com/bahmutov/cache-require-paths

i also wrote up a thing that spins up a persistent server, you can invalidate the cache as you go along: https://github.com/kevinburke/lucifer

i think it's not invalidating things everywhere they need to be, though, which is causing problems when re-running tests.

**Comment by tjwebb** [comment]
Cool, thanks for the info and the follow-up. Yea `require()` already does a lot of caching so it doesn't surprise me that there's not a lot of margin for improvement.

**Comment by mikermcneil** [comment]
@kevinburkeshyp the bottleneck with require is that it's synchronous (like `fs.readFileSync`), so only one require call is being processed at a time.  I'm right there with you -- unfortunately, Isaacs has made [his stance](https://groups.google.com/d/msg/nodejs/y_-LZqltb1A/B1k_o2PRdaMJ) pretty clear on this in the past, and he reiterated the same thing in person when we got a chance to meet up earlier his year.  So I'm afraid we're stuck w/ it as-is for the moment :(

**Comment by mikermcneil** [comment]
@kevinburkeshyp in case it's of interest, I [attempted to optimize](http://mikermcneil.com/post/83604262714/sailsload-latency-latest-after-optimizing) this limitation of `require` elsewhere in core a while back, but eventually abandoned that effort because it wasn't adding a substantial enough benefit to justify how much harder it made it to work with the core code base.

So as you probably figured, there is... one way around the require thing in includeall.  Back in early 2013, I rewrote a lot of Sails core in advance of releasing v0.9, which is where the hooks system came from.  Towards the end of that process, I went back and did a bunch of work to make the moduleloader (and the code that calls it in other hooks) asynchronous, with the hopes that at some point we could implement our own async require logic (or better yet, use an existing implementation).

includeall is actually just a fork I made of Felix's requireall module from a long time ago (which is why it doesn't match the coding style used elsewhere in core)  At that time, while I knew it was possible, I didn't completely understand what would be necessary to replace it.  Since things have been more or less ok (although I'd like to see those 2.5 seconds be more like 0.5 seconds too), it hasn't come up.  But looking back at it now, it's pretty clear that something like this would need to happen:
+ use `async.eachLimit` to loop over the array of targets and `fs.readFile` them
+ For each file:
  + build a code string that looks something like `var ____someReallyWeirdVarName; var module={exports: {}};  ____someReallyWeirdVarName = module;\n`.
  + concatenate code that implements a fake `require` function that calls out to the normal `require()`, but as if it was running from the require context of the path from whence the file was read.  This is the hardest part, imo.  I took a stab at it in the machine runner a while ago and it's not trivial ([these docs are helpful](https://nodejs.org/api/modules.html)).  However, it should be a bit easier to implement when talking about a single app (rather than apps nested inside of apps, etc)  from whence the file was read particular directory.
  + Now concatenate the file contents onto our code string.
  + and `eval()` it.
  + Now `____someReallyWeirdVarName` is the module, as if it had been required the old-fashioned way.

>  i'll probably see if i can work on a PR for this.
I'd love to merge a PR that does this!

**Comment by kevinburkeshyp** [comment]
Yeah agreed. Best option at this point I think might be something that
doesn't require you to load all of sails to e.g. test a single model, but
not sure how much time I'll have to investigate that

-kevin

On Sep 21, 2015, at 16:38, Mike McNeil <notifications@github.com> wrote:

@kevinburkeshyp <https://github.com/kevinburkeshyp> the bottleneck with
require is that it's synchronous (like fs.readFileSync), so only one
require call is being processed at a time. I'm right there with you --
unfortunately, Isaacs has made his stance
<https://groups.google.com/d/msg/nodejs/y_-LZqltb1A/B1k_o2PRdaMJ> pretty
clear on this in the past, and he reiterated the same thing in person when
we got a chance to meet up earlier his year. So I'm afraid we're stuck w/
it as-is for the moment :(

—
Reply to this email directly or view it on GitHub
<https://github.com/balderdashy/include-all/issues/10#issuecomment-142138367>
.

**Comment by sailsbot** [comment]
Thanks for posting, @kevinburkeshyp.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
